Duplicate metrics should not lead to failure

ambroff opened the pull request:
There have been many bugs that have crept in over the years that lead to
some sort of broker failure because sensors were not cleaned up properly
during some normal state transition.

KAFKA-8066 is a good example of this. Even after the fix in 6ca899e for
that bug, we are still seeing the same problem in production.

Rather than trying to find all possible code paths that can lead to
Selector#close not being invoked, this patch just makes duplicate
metrics log an error without throwing.

While duplicate metrics from multiple sensors can be the sign of another
bug, making it fatal seems like a mistake. A common failure we see is
that a replica fetcher thread does not shut down cleanly, leaving behind
some metrics as the thread dies (taking its NetworkClient with
it). Later, when the broker tries to start a replacement replica fetcher
thread in response to a LeaderAndIsr request from the controller, the
fetcher thread cannot start due to this sensor conflict.

Failing to start a replica fetcher in this case can lead to a data loss
situation, since we could end up with the cluster in a state where we
cannot safely restart any brokers to clear up this situation without
offline partitions.

Rather than continuing this trend, we should make sensor duplication
non-fatal.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ambroff committed:
Duplicate metrics should not lead to failure

There have been many bugs that have crept in over the years that lead to
some sort of broker failure because sensors were not cleaned up properly
during some normal state transition.

KAFKA-8066 is a good example of this. Even after the fix in 6ca899e for
that bug, we are still seeing the same problem in production.

Rather than trying to find all possible code paths that can lead to
Selector#close not being invoked, this patch just makes duplicate
metrics log an error without throwing.

While duplicate metrics from multiple sensors can be the sign of another
bug, making it fatal seems like a mistake. A common failure we see is
that a replica fetcher thread does not shut down cleanly, leaving behind
some metrics as the thread dies (taking its NetworkClient with
it). Later, when the broker tries to start a replacement replica fetcher
thread in response to a LeaderAndIsr request from the controller, the
fetcher thread cannot start due to this sensor conflict.

Failing to start a replica fetcher in this case can lead to a data loss
situation, since we could end up with the cluster in a state where we
cannot safely restart any brokers to clear up this situation without
offline partitions.

Rather than continuing this trend, we should make sensor duplication
non-fatal.

radai-rosenblatt commented on clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java:
```diff
@@ -555,8 +555,19 @@ public synchronized void removeReporter(MetricsReporter reporter) {
+
+        KafkaMetric existingMetric = this.metrics.get(metricName);
+        if (existingMetric != null) {
+            log.error("The metric " + metricName + " is being replaced since it had already been registered.");
```
i'd add "please file this as a bug report"

ambroff replied:
+1

radai-rosenblatt commented on clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java:
```diff
@@ -555,8 +555,19 @@ public synchronized void removeReporter(MetricsReporter reporter) {
+                try {
+                    reporter.metricRemoval(existingMetric);
+                } catch (Exception e) {
+                    log.error("Error when removing metric from " + reporter.getClass().getName(), e);
```
could add metric name to this log line

ambroff replied:
Good point, added.

radai-rosenblatt commented on clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java:
```diff
@@ -555,8 +555,19 @@ public synchronized void removeReporter(MetricsReporter reporter) {
-            throw new IllegalArgumentException("A metric named '" + metricName + "' already exists, can't register another one.");
+
+        KafkaMetric existingMetric = this.metrics.get(metricName);
+        if (existingMetric != null) {
```
you may want to make this behaviour controllable (at least as a static boolean) so that testing code can continue to explode.

ambroff replied:
That's a pretty good idea. I ended up doing that as you described with a static boolean, since we apparently have several Metrics repositories in the process, and I think we want this to be a process-wide policy.

The default is to not replace on duplicate, so an exception will be thrown in tests.

radai-rosenblatt commented on clients/src/test/java/org/apache/kafka/common/metrics/MetricsTest.java:
```diff
@@ -377,10 +377,20 @@ public void testSampledStatInitialValue() {
     }
 
-    @Test(expected = IllegalArgumentException.class)
+    @Test
```
if you choose to make this controllable, add a test where this explodes

ambroff replied:
Added.

radai-rosenblatt commented on clients/src/test/java/org/apache/kafka/common/metrics/SensorTest.java:
```diff
@@ -118,14 +117,9 @@ public void testIdempotentAdd() {
 
 
-        // but adding the same metric to a DIFFERENT sensor is an error
+        // but adding the same metric to a DIFFERENT sensor replace that metric with an error in the logs
```
"replace*s* that metric"

ambroff replied:
Fixed.

ambroff committed:
Feedback from Radai

* Make the "replace on duplicate" behavior in the metrics code optional,
  so that it will still explode in tests. Because we have many Metrics
  repositories in the broker (4 by my count in a heap dump I'm looking
  at) we're just opting to make this a static boolean and making this a
  process-wide policy.
* Update some command and log messages based on feedback.

jonlee2 commented on clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java:
```diff
@@ -555,8 +557,23 @@ public synchronized void removeReporter(MetricsReporter reporter) {
+        KafkaMetric existingMetric = this.metrics.get(metricName);
+        if (existingMetric != null) {
+            if (replaceOnDuplicate) {
+                log.error("The metric " + metricName + " is being replaced since it had already been registered. Please file a bug report.");
```
Why not simply calling removeMetric()? Looks like it's doing almost same thing?

ambroff replied:
I guess just to avoid doing a lookup in the map again, but that's harmless, and you're right the code is the same.

I'll do that.

ambroff replied:
Done.

ambroff committed:
Feedback from Jon

Just use removeMetric() before replacing a metric, which does the same
thing.

ambroff committed:
Radai pointed out that we should probably just make this configurable

This has two added benefits:
1.) If someone runs into a bug like this in client code, they can use this same property
2.) Embedded brokers used in integration tests will still default to crashing on duplicate
    sensors, so integration tests will catch silly bugs.

radai-rosenblatt reviewed (approved):
exciting!
